### DISCUSSION TITLE
Different migration result between emulator and smartphone

### DISCUSSION THREAD

**Discussion by mawilms** [simolus3/drift#1644] [Q&A]
Hello everyone,

I'm using Moor for my database interaction and I have this weird behavior where a table isn't created on my actual smartphone while the table is created on my emulator.

My setup:

- Clean app installation on my local smartphone where I have deleted all data that is related to my app
- Tried it in debug and release mode on my smartphone
- Completely wiped emulator
- I'm using on both machines Android 8.0

This is how I create the database:

```dart
LazyDatabase openConnection() {
  return LazyDatabase(() async {
    final dbFolder = await getApplicationDocumentsDirectory();
    final file = File(p.join(dbFolder.path, 'my_database.sqlite'));
    return NativeDatabase(
      file,
      logStatements: true,
    );
  });
}
```

My migration logic is here:

```dart
@override
  MigrationStrategy get migration => MigrationStrategy(
        onCreate: (Migrator m) async {
          await m.createAll();

          if (!Platform.environment.containsKey('FLUTTER_TEST')) {
            await prepopulateDatabase();
          }
        },
      );
```

The table I'm talking about is this one:

```dart
class ExerciseStatistics extends Table {
  IntColumn get fkExercise => integer()();

  IntColumn get fkStatistic => integer()();

  @override
  Set<Column> get primaryKey => {fkExercise, fkStatistic};

  @override
  List<String> get customConstraints => [
    "FOREIGN KEY(fk_exercise) REFERENCES exercises(id) ON DELETE CASCADE ON UPDATE CASCADE",
    "FOREIGN KEY(fk_statistic) REFERENCES statistics(id) ON DELETE CASCADE ON UPDATE CASCADE"
  ];
}
```

It's interesting that the debug log statements with the "CREATE TABLE ..." instructions are only shown in the console when I run the app on the emulator, not on the smartphone.

I have 9 tables. The table listed above is the only one missing. I've double checked the .sqlite files in the "Device File Explorer" and the the emulator file contains 9 tables, the smartphone file 8.

I really appreciate your help :)

**Comment by simolus3** [comment]
Some Android phones like to persist application data even when you uninstall the app. Can you manually wipe your app's data in the settings? Surely after that we should see all the "CREATE TABLE" statements, maybe that helps to see what's going on here.
